spam/spam_tutorial.py

```diff
@@ -41,151 +41,140 @@
+#
+# Because their only requirement is that they map a data point a label (or abstain), they can wrap a wide variety of forms of supervision. Examples include, but are not limited to:
+# * Keyword searches: looking for specific words in a sentence
+# * Pattern matching: looking for specific syntactical patterns
+# * Third-party models: using an pre-trained model (usually a model for a different task than the one at hand)
+# * Distant supervision: using external knowledge base
+# * Crowdworker labels: treating each crowdworker as a black-box function that assigns labels to subsets of the data
+#
+# The process of **developing LFs** is iterative and usually consists of:
+# * Writing a function
+# * Estimating its performance by looking at labeled examples in the training set or dev set
+# * Iterating to improve coverage or accuracy as necessary.
+# Balancing accuracy and coverage for specific labeling functions as well as the overall set of LFs developed is often a trade-off, and depending on the use case, users may tend to prefer one over the other.
+#
+# Once multiple LFs have been created, users can look at data points that have received no labels so far (or many conflicting labels) to get ideas for new LFs to write.
+# Another thing to check is whether there are any classes for which there are very few LFs, so that additional LFs can be written to cover that part of the dataset.
+# **Note, however, that it is not necessary for LFs to assign labels to every data point;** in fact, most of the time your LFs will not have perfect dataset-wide coverage.
+# We rely on the fact that the classifier that trains on labels from Snorkel has the power to _generalize_ and can therefore learn a good representation of the data even if the each data point in the training set does not receive a label from any LFs.
+#
+#
 
 # %% [markdown]
 # ### a) Look at examples for ideas
 
 # %% [markdown]
-# * Look at 10 examples; got any ideas?
+# We begin the process of writing LFs by looking at some examples in the dev set.
 
 # %%
 # Display just the text and label
-df_dev[["CONTENT", "LABEL"]].sample(10, random_state=123)
-
-# %%
-# for i, x in df_dev.iterrows():
-#     if "please" in x.CONTENT:
-#         print(x.CONTENT)
+df_dev[["text", "label"]].sample(10, random_state=123)
 
 # %% [markdown]
 # ### b) Write an LF
 
 # %% [markdown]
-# The simplest way to create labeling functions in Snorkel is with the `@labeling_function()` decorator, which wraps a function for evaluating on a single `DataPoint` (in this case, a row of the dataframe).
+# The recommended way to create labeling functions in Snorkel is with the `@labeling_function()` decorator, which wraps a function for evaluating on a single data point (in this case, a row of the dataframe).
```

Comment:
isn't this mandatory, not recommended? 

_To create labeling functions in Snorkel..._

Reply:
I don't think it technically is (since the decorator just helps with initializing the `LabelingFunction` class), but it's the only supported way, so yeah, I'll make that stronger.